File: server/app/model/database/shop-item.ts
import { removeUnderscoreId } from '@app/utils/transform-utils';
import { ShopItemType } from '@common/enum/shop-item-type';
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { ApiProperty } from '@nestjs/swagger';
import { Document } from 'mongoose';

export type ShopDocument = ShopItem & Document;

@Schema({
    toJSON: {
        virtuals: true,
        versionKey: false,
        transform: removeUnderscoreId,
    },
})
export class ShopItem {
    @ApiProperty({ name: 'id' })
    _id?: string;

    @ApiProperty()
    @Prop({ required: true })
    name: string;

    @ApiProperty()
    @Prop({ required: true })
    type: ShopItemType;

    @ApiProperty()
    @Prop({ required: true })
    cost: number;

    @ApiProperty()
    @Prop({ required: false })
    imageUrl: string;

    @ApiProperty()
    @Prop({ required: false })
    soundUrl: string;
}

export const shopItemSchema = SchemaFactory.createForClass(ShopItem);

File: mobile_client/lib/shop/classes/shop_item.dart
import 'dart:core';

import 'package:mobile_client/shop/enum/shop_item_type.dart';

class ShopItem {
  final String id;
  final String name;
  final ShopItemType type;
  final int cost;
  final String imageUrl;
  final String? soundUrl;

  ShopItem({
    required this.id,
    required this.name,
    required this.type,
    required this.cost,
    required this.imageUrl,
    this.soundUrl,
  });

  factory ShopItem.fromJson(Map<String, dynamic> json) {
    return ShopItem(
      id: json['id'],
      name: json['name'],
      type: ShopItemType.values
          .firstWhere((e) => e.toString().split('.').last == json['type']),
      cost: json['cost'],
      imageUrl: json['imageUrl'] ?? '',
      soundUrl: json['soundUrl'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type.name,
      'cost': cost,
      'imageUrl': imageUrl,
      'soundUrl': soundUrl,
    };
  }
}
